lcls_tools/common/data_analysis/bmad_modeling/bmad_modeling.py

```diff
@@ -0,0 +1,387 @@
+        x_np = np.where(x_np < ampl_reject, 0, x_np)
+    if val_type == 'Phas':
+        x_np = np.deg2rad(x_np)
+    return x_np
+
+
+def bmag(twiss, twiss_reference):
+    """Calculates BMAG from imput twiss and reference twiss"""
+    beta_a, alpha_a, beta_b, alpha_b = twiss
+    beta_a_ref, alpha_a_ref, beta_b_ref, alpha_b_ref = twiss_reference
+    bmag_a = bmag_func(beta_a, alpha_a, beta_a_ref, alpha_a_ref)
+    bmag_b = bmag_func(beta_b, alpha_b, beta_b_ref, alpha_b_ref)
+    return (bmag_a, bmag_b)
+
+
+def bmag_func(bb, ab, bl, al):
+    """bmag function"""
+    return 1 / 2 * (bl / bb + bb / bl + bb * bl * (ab / bb - al / bl) ** 2)
+
+
+def get_bmad_bdes(tao, element, b1_gradient=[]):
+    """Returns BDES from Bmad B1_GRADIENT or given gradient"""
+    ele_attr = tao.ele_gen_attribs(element)
+    if not b1_gradient:
+        b1_gradient = ele_attr["B1_GRADIENT"]
+    return -b1_gradient * ele_attr["L"] * 10
+
+
+def match_twiss(tao, variable, datum):
+    """Performs optimization for given variable and datum"""
+    match_cmds = ["veto var *", "veto dat *@*",
+                  "set global n_opti_cycles = 912"]
+    match_cmds.append(f"use var {variable}[1:4]")
+    match_cmds.append(f"use dat {datum}[1:4]")
+    tao.cmds(match_cmds)
+    tao.cmd("run")
+
+
+def get_tao(pvdata, mdl_obj):
+    """
+    Returns tao commands from pvdata, if data_source is DES, calls use_klys_when_beam off for Cu Linac
+    """
+    lines_quads, lines_rf = [], []
+    for dm_key, map in mdl_obj.all_data_maps.items():
+        if dm_key.startswith("K"):
+            acc_pv = map.accelerate_pvname
+            if acc_pv == "":
+                continue
+            # acc_pv = f"{acc_pv[0:21]}{mdl_obj.beam_code}{acc_pv[22:]}"
+            # map.accelerate_pvname = acc_pv
+            lines_rf += map.as_tao(pvdata)
+        if dm_key == "cavities":
+            lines_rf = map.as_tao(pvdata)
+        if dm_key == "quads":
+            lines_quads += map.as_tao(pvdata)
+    if "NOT" in mdl_obj.data_source and "cu" in mdl_obj.beam_path:
+        new_lines = []
+        for cmd in lines_rf:
+            if "in_use" in cmd:
+                if "K21_1" in cmd or "K21_2" in cmd:
+                    new_lines.append(cmd)  # L1 always on Beam Code 1
+                    continue
+                ele = cmd.split()[2]
+                [sector, station] = ele[1:].split("_")
+                pv = f"KLYS:LI{sector}:{station}1:"
+                f"BEAMCODE{mdl_obj.beam_code}_STAT"
+                cmd_words = cmd.split()
+                cmd_words[-1] = str(pvdata[pv])
+                new_lines.append(" ".join(cmd_words))
+            else:
+                new_lines.append(cmd)
+        # lines_rf = new_lines
+    return lines_rf + lines_quads
+
+
+def get_machine_values(data_source, pv_list, date_time=''):
+    """Returns pvdata, a dictionary with keys containing the PV name and values from Actual, Desired or Archive"""
+    if data_source in ["DES", "ACT"]:
```

Comment:
It would be nice to throw an error that reminds the user what allowable data_source values are. The code as written will error out if pvdata is not defined.

Reply:
add message with valid values of data_source